src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestEventsConverter.kt

```diff
@@ -0,0 +1,203 @@
+import com.intellij.execution.testframework.sm.ServiceMessageBuilder
+import com.intellij.execution.testframework.sm.runner.OutputToGeneralTestEventsConverter
+import com.intellij.openapi.util.Key
+import jetbrains.buildServer.messages.serviceMessages.ServiceMessage
+import jetbrains.buildServer.messages.serviceMessages.ServiceMessageVisitor
+
+/**
+ * Translate Cargo Test output to TeamCity Service Messages which SM Test Runner understands.
+ *
+ * Unfortunately, as for now Cargo Test only supports `human` message format, which is pretty
+ * obscure to handle programmatically. First of all, it is split into some separate sections,
+ * which are not easily distinguishable, that's why we do state machines here. Next problem is,
+ * that Cargo collects failed tests stdouts and print them yet at the end of test session.
+ * Lastly, we do not have exact information when given test started.
+ *
+ * For SM docs see: [https://confluence.jetbrains.com/display/TCD10/Build+Script+Interaction+with+TeamCity].
+ *
+ * Some parts based on [TeamCity Rust Plugin](https://github.com/JetBrains/teamcity-rust-plugin).
+ */
+class CargoTestEventsConverter(consoleProperties: TestConsoleProperties) :
+    OutputToGeneralTestEventsConverter("Cargo Test", consoleProperties) {
+
+    private val failedTests: MutableMap<String, FailedTestData> = mutableMapOf()
+    private var currentTestName: String? = null
+
+    private var visitor: ServiceMessageVisitor? = null
+
+    private var _state: State = State.Building
+    private var state: State
+        get() = _state
+        set(value) {
+            _state.onLeave(this)
+            _state = value
+            _state.onEnter(this)
+        }
+
+    override fun dispose() {
+        check(state == State.Summary)
+        state.onLeave(this)
+        super.dispose()
+    }
+
+    override fun processServiceMessages(text: String, outputType: Key<*>, visitor: ServiceMessageVisitor): Boolean {
+        if (this.visitor == null) {
+            this.visitor = visitor
+            state.onEnter(this)
+        }
+
+        val cleanText = text.trim()
+
+        val currentState = state
+        val result = state.onMessage(this, cleanText, outputType)
+        if (!result && currentState != state) {
+            // If we have changed state and haven't processed
+            // current message, try to pass it to new state.
+            return state.onMessage(this, cleanText, outputType)
+        } else {
+            return result
+        }
+    }
+
+    private fun message(sm: String) {
+        val message = requireNotNull(ServiceMessage.parse(sm)) { "message processor should build valid SMs" }
+        message.visit(requireNotNull(visitor) { "visitor hasn't be initialized" })
+    }
+
+    private fun message(smb: ServiceMessageBuilder) = message(smb.toString())
+
+    // Kotlin chokes on inner sealed classes, that's why we use enums here
+    private enum class State {
```

Comment:
Ah, reminds me of good ol' days of jps: https://github.com/intellij-rust/intellij-rust/blob/239096bfc0d486acac17e1669ed9b34e4b665241/jps/src/main/kotlin/org/rust/jps/RustBuilder.kt#L123